### PR TITLE
✨ Start drafting placement translator internal interfaces

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
## Summary
This PR is the first draft of the internal interfaces among the parts of the placement translator process of the 2023q1 PoC.

This draft assumes that #155 will be resolved by switching to using the SyncTarget UID instead of its name in the construction of the mailbox workspace name.

## Related issue(s)

Fixes #

### CODE REVIEW

File: cmd/placement-translator/main.go [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+	inventoryConfig.UserAgent = "mailbox-controller"
+
+	// Get client config for view of SyncTarget objects
+	edgeViewConfig, err := configForViewOfExport(ctx, restConfig, "edge.kcp.io")
+	if err != nil {
+		logger.Error(err, "Failed to create client config for view of edge APIExport")
+		os.Exit(4)
+	}
+
+	edgelusterClientset, err := edgeclusterclientset.NewForConfig(edgeViewConfig)
```
Comment (waltforme) [inline cmd/placement-translator/main.go:148]: edge**C**lusterClientset?

File: pkg/placement/declarations.go [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+// The implementation may atomize the resolved "where" using a SinglePlacementSliceSetReducer.
+// The implementation may use a BindingOrganizer to get from the atomized
+// "what" and "where" to the ProjectionMapProducer behavior.
+type SetBinder interface {
+	AsWhatConsumer() DynamicMapConsumer[edgeapi.ExternalName, WorkloadParts]
+	AsWhereConsumer() DynamicMapConsumer[edgeapi.ExternalName, ResolvedWhere]
+	ProjectionMapProducer
+}
+
+// WorkloadProjector is kept appraised of what goes where
```
Comment (waltforme) [inline pkg/placement/declarations.go:83]: Is WorkloadProjector kept appraised of 'what goes where' or just 'what'?

File: pkg/placement/declarations.go [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+// ProjectionKey identifies the topmost level of organization,
+// the combinatin of the destination and the API group and resource.
+type ProjectionKey struct {
+	metav1.GroupResource
+	Destination SinglePlacement
+}
+
+// ProjectionPerCluster is the second level of organization.
+// It identifies the API version to use currently and holds
+// the map provider that gets to the lowest level of organization.
```
Comment (waltforme) [inline pkg/placement/declarations.go:182]: I found 'map provider' and 'map producer' are both used here and it seems that two terms refer to the same thing. We may want to be consistent for easier reading.

File: pkg/placement/declarations.go [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+	// For non-namespaced objects, Names can optionally be non-nil to restrict
+	// the objects handled.
+	Names *k8ssets.String
+}
+
+// SinglePlacement extends the API struct with the UID of the SyncTarget.
+type SinglePlacement struct {
+	edgeapi.SinglePlacement
+	SyncTargetUID apimachtypes.UID
```
Comment (waltforme) [inline pkg/placement/declarations.go:209]: We may need to change this because now #198 is merged.

It is just a kind reminder in case it is omitted, because it is easier to be forgotten comparing to other explicit usage of `ExternalName`.

File: pkg/placement/map-types.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package placement
+
+// DynamicMapProducer holds a mutable map and keeps consumers appraised of it.
+// The zero value of Val signals a missing entry in the map.
+type DynamicMapProducer[Key comparable, Val any] interface {
```
Comment (waltforme) [inline pkg/placement/map-types.go:21]: Question: It is 'dynamic' because it can produce values in continuous and incremental manner, right?

File: pkg/placement/map-types.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	// The producer precedes each consumer in the locking order.
+	AddConsumer(consumer DynamicMapConsumer[Key, Val], notifyCurrent bool)
+
+	// Get invokes the given function on the value corresponding to the key.
+	// This does not count as informing any particular consumer.
+	// The producer precedes the function in the locking order.
+	Get(Key, func(Val))
+}
+
+// DynamicMapProducerGet does a non-CPS Get.
```
Comment (waltforme) [inline pkg/placement/map-types.go:36]: Question: What is 'non-CPS'?

File: pkg/placement/map-types.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	Set(Key, Val)
+}
+
+// DynamicMapConsumerFunc is a func value that implements DynamicMapConsumer.
+// Remember that func values are not comparable.
+type DynamicMapConsumerFunc[Key comparable, Val any] func(Key, Val)
+
+func (cf DynamicMapConsumerFunc[Key, Val]) Set(key Key, val Val) { cf(key, val) }
+
+type Client[T any] interface {
```
Comment (waltforme) [inline pkg/placement/map-types.go:71]: How about `Consumer`? To resonate with other names.

**Review by waltforme — COMMENTED** [review]
I'm learning from this PR which designs a generic event-driven data flow mechanism IMO.

File: pkg/placement/declarations.go (hunk 2 above)
Comment (MikeSpreitzer) [reply pkg/placement/declarations.go:83]: 'what goes where'

File: pkg/placement/declarations.go (hunk 4 above)
Comment (MikeSpreitzer) [reply pkg/placement/declarations.go:209]: Yes, I will rebase this PR.

File: pkg/placement/map-types.go (hunk 6 above)
Comment (MikeSpreitzer) [reply pkg/placement/map-types.go:36]: "continuation-passing style"

File: pkg/placement/map-types.go (hunk 7 above)
Comment (MikeSpreitzer) [reply pkg/placement/map-types.go:71]: A "Consumer" is generally invoked multiple times, a client only needs to be given its provider once.

Comment (MikeSpreitzer) [reply pkg/placement/map-types.go:83]: I adjusted the terminology further in the latest push.

File: pkg/placement/declarations.go (hunk 3 above)
Comment (MikeSpreitzer) [reply pkg/placement/declarations.go:182]: In the latest push I tried to rationalize the terminology.

**Comment by MikeSpreitzer** [comment]
The force-push to 5fb9e73 removes an obsolete reference UIDer.

File: pkg/placement/test_uider.go [hunk 8]
```diff
@@ -51,7 +51,7 @@ func NewTestUIDer(rng *rand.Rand, wg *sync.WaitGroup) *testUIDer {
 	}
 }
 
-func (tu *testUIDer) AddConsumer(consumer DynamicMapConsumer[ExternalName, apimachtypes.UID], notifyCurrent bool) {
+func (tu *testUIDer) AddReceiver(consumer MappingReceiver[ExternalName, apimachtypes.UID], notifyCurrent bool) {
```
Comment (waltforme) [inline pkg/placement/test_uider.go:54]: The first parameter of `AddReceiver` seems to be passive in this situation because its `Set` is called by `tu`, not by itself. Also the doc of `MappingReceiver` says 'MappingReceiver is given map entries by a DynamicMapProvider.'

OTOH `tu` implements `DynamicMapProvider`, so by the updated naming convention, it passively provides the service of a DynamicMap.

So both `tu` and `consumer` are passive? I'm trying hard to follow.

Comment (MikeSpreitzer) [reply pkg/placement/test_uider.go]: I made another revision, hopefully it is clearer now.

File: cmd/placement-translator/main.go (hunk 1 above)
Comment (MikeSpreitzer) [reply cmd/placement-translator/main.go]: Fixed in latest push.
